CONTRIBUTING.md

```diff
@@ -0,0 +1,853 @@
+
+### Indentation and vertical whitespace
+
+Indent with hard tabs. Each one counts for 4 spaces.
+
+Never mix tabs and spaces. Don't use smart tabs.
+
+### Excessive whitespace
+
+Keep whitespace within reason. Limit vertical whitespace to no more than 2 consecutive blank lines, and don't start or end blocks with plain whitespace. Don't use more than one character of horizontal whitespace beyond indentation.
+
+```js
+// Bad
+function iterate() {
+	var index = 0
+	var list = []
+
+	while (hasNext()) {
+		list.push(next(index))
+		index++
+	}
+
+	// Too many empty lines
+
+
+
+	for (var i = 0; i < list.length; i++) {
+		read(list[i])
+	}
+}
+
+// Also bad
+var a  =  2
+//   ^^ ^^
+```
+
+### Control keywords
+
+Always surround control keywords (e.g. `if`, `else`, `for`) with whitespace.
+
+## Operators
+
+### Binary operators
+
+Always surround binary keyword operators (e.g. `in`, `instanceof`) with whitespace.
+
+Always surround any other binary operator with whitespace, including assignment operators. Add line breaks after the operator, not before.
+
+```js
+// Good
+var a = 1 + 2
+a = 3
+a += 4
+
+var a = 1 + 2 + 3 + 4 + 5 +
+	6 + 7 + 8 + 9 + 10
+
+// Bad
+var a = 1+2
+var a = 1 +2
+var a = 1+ 2
+var a=1+2
+var a=1 + 2
+a=3
+a+=4
+
+var a = 1 + 2 + 3 + 4 + 5
+	+ 6 + 7 + 8 + 9 + 10
+```
+
+As an exception, casting to a 32-bit integer (i.e. `x|0`) doesn't require whitespace surrounding it.
+
+```js
+// This is okay
+var casted = value|0
+```
+
+### Unary operators
+
+Always use whitespace between an unary keyword operator (e.g. `delete`, `new`) with whitespace.
+
+Do not use spaces between any other unary operator and the value they apply to.
+
+```js
+// Good
+!a
+!!a
+~~a
+a++
+++a
+-1
+~1
+
+// Okay
++!a
+
+// Bad
+! a
+!! a
+! ! a
+++ a
+a ++
++ ! a
+
+// Even worse
+- 1
+~ 1
+```
+
+Don't use `new function () {}`. Just use an object literal instead, and an IIFE if needed.
+
+## Objects
+
+### Exterior whitespace
+
+Do not include space between the opening curly brace and the first object key in single line objects. For multi-line objects, don't include any extra preceding whitespace at the beginning or end.
+
+```js
+// Good
+var object = {foo: 1}
+
+var object = {
+	foo: 1,
+	bar: 2,
+	baz: 3,
+	quux: "string"
+}
+
+// Bad
+var object = { foo: 1 }
+
+var object = {
+
+	foo: 1,
+	bar: 2,
+	baz: 3,
+	quux: "string"
+
+}
+```
+
+### Interior whitespace
+
+Use no space before an object key and the colon, but use a single space between the colon and value.
+
+```js
+// Good
+var object = {foo: 1}
+
+// Bad
+var object = {foo:1}
+var object = {foo : 1}
+
+// Very bad
+var object = {foo :1}
+```
+
+### Larger objects
+
+Non-trivial objects should occupy multiple lines.
+
+```js
+// Good
+var object = {
+	foo: 1,
+	bar: 2,
+	baz: 3,
+	quux: "string"
+}
+
+// Bad
+var object = {foo: 1, bar: 2, baz: 3, quux: "string"}
+```
+
+### Comma placement
+
+Commas come last. Also, don't use trailing commas.
```

Comment:
Okay. I normally include them myself. Just the existing style didn't
usually. I tried to limit changes to what already existed before. I'll
change everything in the project when I get home.

On Mon, Nov 9, 2015, 15:04 Richard Eames <notifications@github.com> wrote:

> In CONTRIBUTING.md
> <https://github.com/lhorie/mithril.js/pull/834#discussion_r44323525>:
>
> > +```js
> > +// Good
> > +var object = {
> > +	foo: 1,
> > +	bar: 2,
> > +	baz: 3,
> > +	quux: "string"
> > +}
> > +
> > +// Bad
> > +var object = {foo: 1, bar: 2, baz: 3, quux: "string"}
> > +```
> > +
> > +### Comma placement
> > +
> > +Commas come last. Also, don't use trailing commas.
>
> Why not use trailing commas? In the whitespace section you have
>
> Please don't leave trailing spaces anywhere. It makes diffs harder to read.
>
> I would use the same argument for *excluding* trailing commas: it makes
> diffs harder to read.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lhorie/mithril.js/pull/834/files#r44323525>.
>